File: tests/test_ops.cpp
#include "gtest/gtest.h"

#include <cstdio>
#include <sstream>
#include <vector>
#include <tuple>
#include <algorithm>

#include "tinymesh/tinymesh.h"
using namespace tinymesh;

#include "test_utils.h"

class MeshOpsTest : public TinyMeshBaseTestWithParam {
public:
    MeshOpsTest() = default;
    virtual ~MeshOpsTest() = default;

    void SetUp() override {
        mesh.load(filename);
        mesh.fillHoles();
    }
};

// Operators

TEST_P(MeshOpsTest, GetLaplacian) {
    ASSERT_NO_FATAL_FAILURE(getMeshLaplacian(mesh, MeshLaplace::Adjacent));
    ASSERT_NO_FATAL_FAILURE(getMeshLaplacian(mesh, MeshLaplace::Cotangent));
    ASSERT_NO_FATAL_FAILURE(getMeshLaplacian(mesh, MeshLaplace::Belkin08));
}

// Features

TEST_P(MeshOpsTest, GetHeatKernelSignatures) {
    auto L = getMeshLaplacian(mesh, MeshLaplace::Cotangent);
    const int K = std::min((int)L.rows(), 200);
    ASSERT_NO_FATAL_FAILURE(getHeatKernelSignatures(L, K));
}

TEST_P(MeshOpsTest, GetPrincipalCurvatures) {
    ASSERT_NO_FATAL_FAILURE(getPrincipalCurvatures(mesh));
}

static std::vector<std::string> filenames = {
    "torus.obj",
    "fandisk.ply",
    "bunny_mini.ply",
};

INSTANTIATE_TEST_SUITE_P(, MeshOpsTest, ::testing::ValuesIn(filenames));

File: tests/python/test_ops.py
import os
from functools import partial
from itertools import product

import pytest

import tinymesh as tms

CWD = os.getcwd()

model_dir = 'data/models'
filenames = [
    'torus.obj',
    'fandisk.ply',
    'bunny_mini.ply',
]

methods = [
    partial(tms.get_mesh_laplacian, type=tms.MeshLaplace.ADJACENT),
    partial(tms.get_mesh_laplacian, type=tms.MeshLaplace.COTANGENT),
    partial(tms.get_mesh_laplacian, type=tms.MeshLaplace.BELKIN08),
    tms.get_principal_curvatures,
    tms.get_principal_curvatures_with_derivatives,
    tms.get_feature_line_field,
]


@pytest.mark.parametrize("filename", filenames)
def test_hks(filename):
    filename = os.path.join(CWD, model_dir, filename)
    mesh = tms.Mesh(filename)
    mesh.fill_holes()

    L = tms.get_mesh_laplacian(mesh, tms.MeshLaplace.COTANGENT)
    K = min(L.shape[0], 200)

    try:
        tms.get_heat_kernel_signatures(L, K)
    except Exception:
        pytest.fail('Failed!')

    assert mesh.verify(), "Mesh verification failed!"


@pytest.mark.parametrize("method, filename", product(methods, filenames))
def test_ops_method(method, filename):
    filename = os.path.join(CWD, model_dir, filename)
    mesh = tms.Mesh(filename)
    mesh.fill_holes()

    try:
        method(mesh)
    except Exception:
        pytest.fail('Failed!')

    assert mesh.verify(), "Mesh verification failed!"
